In [37]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression

In [238]:
#read in clean data
comma = pd.read_csv("C:\\Users\\ja041718\\Downloads\\clean_data.txt", sep="\t", header=0)
comma.columns

Index(['Unnamed: 0', 'RespondentID',
       'Sentence_It's important for a person to be honest, kind, and loyal.',
       'OxfordCommaHeard_Yes', 'OxfordCommaCare_Not at all',
       'OxfordCommaCare_Not much', 'OxfordCommaCare_Some',
       'Unnamed: 4_Some experts say it's important to drink milk, but the data is inconclusive.',
       'DataPlural_Yes', 'DataPluralCare_Not at all',
       'DataPluralCare_Not much', 'DataPluralCare_Some',
       'GrammarImportance_Somewhat important',
       'GrammarImportance_Somewhat unimportant',
       'GrammarImportance_Very important',
       'GrammarImportance_Very unimportant', 'Income_$100,000 - $149,999',
       'Income_$150,000+', 'Income_$25,000 - $49,999',
       'Income_$50,000 - $99,999', 'Education_Graduate degree',
       'Education_High school degree',
       'Education_Less than high school degree',
       'Education_Some college or Associate degree',
       'Location_East South Central', 'Location_Middle Atlantic',
       'Location

We need to setup the indexes for the columns used in the regression. Our dependent variable is in column 32, but we also need to remove columns 4-7 since they represent the same survey question.

In [257]:
#Indexes of predictors
predcols = list(range(3,33))

#Dependent Variable
labelcol = 32

#Variables related to the dependent variable that need to be removed
sublabels = [5,6,4]

#Remove the variables that need to be removed from the predictors
predcols.remove(labelcol)

for col in sublabels:
    predcols.remove(col)
    
print(predcols)

[3, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]


In [258]:
#take out the dependent variable
labels = comma.iloc[:,labelcol]



#take out the independent variables
predictors = comma.iloc[:,  predcols]

In [191]:
#Setup a logistic regression object
glm = LogisticRegression()

In [259]:
#fit the glm model
commaglm = glm.fit(predictors, labels)
commaglm_prob = commaglm.predict_proba(predictors)
df = pd.DataFrame(commaglm_prob)

In [260]:
#Score of the model
commaglm.score(predictors, labels)

0.76759884281581481

In [261]:
#Actual rate of people who don't think the comma is important
1 - labels.mean()

0.7299903567984571

So we can see our model does well at 76.8%, but you could achieve 73% just by guessing that everyone doesn't think commas are important. Therefore the model is not that great.

In [262]:
#append the list of all coefficients used in the model
coef=[]
for i in commaglm.coef_:
    coef.append(i)

In [263]:
#create a df of the coefficients with their labels
df = pd.DataFrame(list(comma.columns[predcols]),
                  coef)
df

0
 1.127107                               OxfordCommaHeard_Yes
-0.081100  Unnamed: 4_Some experts say it's important to ...
 0.082882                                     DataPlural_Yes
-1.526581                          DataPluralCare_Not at all
-1.408221                            DataPluralCare_Not much
-0.881492                                DataPluralCare_Some
-0.303399               GrammarImportance_Somewhat important
 0.193160             GrammarImportance_Somewhat unimportant
 1.135965                   GrammarImportance_Very important
-0.392956                 GrammarImportance_Very unimportant
-0.245321                         Income_$100,000 - $149,999
-0.433458                                   Income_$150,000+
-0.417222                           Income_$25,000 - $49,999
-0.794780                           Income_$50,000 - $99,999
-0.144969                          Education_Graduate degree
-0.131899                       Education_High school degree
 0.554773             Education_Less than high school degree
-0.146236         Education_Some college or Associate degree
 0.039258                        Location_East South Central
-0.238292                           Location_Middle Atlantic
-0.317563                                  Location_Mountain
 0.283532                               Location_New England
-0.210601                                   Location_Pacific
 0.312409                            Location_South Atlantic
 0.083301                        Location_West North Central
-0.201768                        Location_West South Central